server/service/listing.js

```diff
@@ -0,0 +1,182 @@
+const geolib = require('geolib');
+const stores = require("./stores");
+
+const NUMBER_OF_BUCKETS = 5;
+
+async function findStoreListings(params, debug = false){
+    let listings = await stores.findNearbyStores(params);
+    addScoringParameters(listings, params);
+    let listingsByBucket = groupByBucket(listings);
+    let sortedListingsByBucket = sortBuckets(listingsByBucket);
+    return flattenListings(sortedListingsByBucket);
+}
+
+function groupByBucket(storeListings){
+    return storeListings.reduce(function(rv, x) {
+        rv[x.scoreParams.bucket] = rv[x.scoreParams.bucket] || {};
+        if(Array.isArray(rv[x.scoreParams.bucket][x.scoreParams.quality])){
```

Comment:
this feels a bit prone to a potential undefined exception. would it help to initialize with a complete object (with zero values) instead of an empty object?

Reply:
Do you mean `{}` object ? 
Wouldn't `Array.isArray` handle those ?

Reply:
it will handle it for now, and the changes here are all correct. without these handrails though, chances are we'll inadvertently land something that breaks in the future (like it happened with `null` store categories)